### PR TITLE
Midi Support GoDAW

### PR DESCRIPTION
**Pull request by abhinav3967**
Midi Support for GoDAW

### CODE REVIEW

File: util/Midi/MIDI.gd [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+	MetaLyrics = 0x05,
+	MetaMarker = 0x06,
+	MetaCuePoint = 0x07,
+	MetaChannelPrefix = 0x20,
+	MetaEndOfTrack = 0x2F,
+	MetaSetTempo = 0x51,
+	MetaSMPTEOffset = 0x54,
+	MetaTimesSignature = 0x58,
+	MetaKeySignature = 0x59,
+	MetaSequencerSpecifi = 0x7F,
```
Comment (Technohacker) [inline util/Midi/MIDI.gd:62]: Intentionally cut off or a typo?

File: util/Midi/MIDI.gd [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+export var bpm = 0
+
+
+# function to parse files to midi
+func parse_file(filename: String = "") -> bool:
+	# Open file, if any errors return false
+	var input_file = File.new()
+	var err = input_file.open(filename, File.READ)
+	if err:
+		return false
```
Comment (Technohacker) [inline util/Midi/MIDI.gd:78]: Add a push_error here just in case so we know what the error is :)

File: util/Midi/MIDI.gd [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+					tracks[chunk].events.append(MidiEvent.new(MidiEvent.Other))
+				EventName.SystemExclusive:
+					if status == 0xF0:
+						#System Exclusive Message Begin
+						print("System Exclusive Begin: " + read_string(input_file,
+							read_value(input_file)))
+					if status == 0xF7:
+						print("System Exclusive End: " + read_string(input_file,
+							read_value(input_file)))
+					if status == 0xFF:
```
Comment (Technohacker) [inline util/Midi/MIDI.gd:166]: Maybe make this a match, and also make an enum from these constants?

File: util/Midi/MIDI.gd [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+# Swap byte order of integers as midi was made years ago
+# and the byte order have changed over the years.
+
+# Swaps byte order of 32-bit integer
+func swap32(n):
+	return (((n >> 24) & 0xff) | ((n << 8) & 0xff0000) | ((n >> 8) & 0xff00) | ((n << 24) & 0xff000000))
+
+# swaps byte order of 16-bit integer
+func swap16(n):
+	return ((n >> 8) | (n << 8))
```
Comment (Technohacker) [inline util/Midi/MIDI.gd:236]: _Should_ be unnecessary, I'll try using endian_swap

**Review by Technohacker — CHANGES_REQUESTED** [review]

File: util/Midi/MIDI.gd (hunk 1 above)
Comment (abhinav3967) [reply util/Midi/MIDI.gd:62]: typo

File: util/Midi/MIDI.gd (hunk 2 above)
Comment (abhinav3967) [reply util/Midi/MIDI.gd:78]: will do

File: util/Midi/MIDI.gd (hunk 3 above)
Comment (abhinav3967) [reply util/Midi/MIDI.gd:166]: okay

**Comment by abhinav3967** [comment]
Resolved all Errors

**Review by Technohacker — APPROVED** [review]

**Pull request merged**